81-IDevID_and_IAK_support.md

```diff
@@ -0,0 +1,302 @@
+enhancement ecosystem.
+
+How will security be reviewed and by whom?
+-->
+
+
+## Design Details
+
+<!--
+This section should contain enough information that the specifics of your
+change are understandable.  This may include API specs (though not always
+required) or even code snippets.  If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+A potential workflow is presented below for the scenario using EK, EK certificate, IDevID, IAK, LDevID and LAK. The other scenarios will include most parts of this workflow. It also worth mentioning that when using an IAK the EK is not needed to enroll another AK (being either an EAK or an LAK). Instead of using make/activate credential, which requires a decryption key, TPM2_Certify is used to prove that the EAK/LAK (including the sensitive area, e.g., the plaintext private key) is co-resident in the same TPM as the IAK.
+
+### Registration from the agent perspective when interacting with the registrar service.
+1. Create session (done at get_tpm2_ctx)   
+2. Get endorsement certificate from TPM NV index and EK public key (done at create_ek)
+3. Regeneration of IAK and IDevID possible info (create create_iak and create_idevid routines). Depending on the configuration, the IDevID/IAK may also require a session.
+4. Load the IDevID and IAK and their certificates from a given local path and send them along with the rest of the attestation/measured data to the registrar service
+5. Create keys for LAK and LDevID based on the IAK (if possible, load them if already created). When using the TPM2_Certify the nonce should be used here.
+6. Encode the necessary data and send it to the registrar service (modify the method do_register_agent to support the new parameters regarding IDevID, LDevID, IAK and LAK keys). 
+7. Receive the challenge to prove that is actually has the IDevID, LDevID, IAK and LAK keys. 
+8. Sends challenge response back to the registrar service.
+9. The agent gets its UUID in case of a successful registering process.
+
+### Registration from the registrar service perspective when interacting with the agent
+1. The main endpoints here are do_register_agent() and do_activate_agent() invoked by the agent.
+2. Decode data and parse certificates. This implies modifying the method doRegisterAgent to support new parameters regarding IAK and IDevID. This method exercises a POST to /v{api_version}/agents/{agent_id} at registrar_common.py do_POST method.  
+3. Verify IDevID certificate chain of trust + IAK certificate chain of trust
+4. Create a challenge to send to the agent for prove the possession of the private keys. It's important to mention that the challenge will change from using Make/ActivateCredential to use TPM2_Certify.
+4. Create a challenge to send to the agent for prove the possession of the private keys. 
+5. Receive the challenge response and process
+6. Generate the certificates for LAK and LDevID keys with the owner/company's CAs. 
```

Comment:
Here the generation of the LAK is part of the protocol. We should define then on how this should actually work in Keylime.

Reply:
Makes sense!